### ISSUE TITLE
Additional false positives in `PLR1733` and `PLR1736`

### ISSUE THREAD

**Issue by ntBre** [astral-sh/ruff#25182]
This is a follow-up issue to https://github.com/astral-sh/ruff/issues/25150, tracking a couple of additional false positives. https://github.com/astral-sh/ruff/pull/25177 fixed the first one:

```py
def foo(d: dict):
    for k, v in d.items():
        ...
    else:
        print(d[k])  # resolved false positive
```

This was a false positive because `k` may be unbound in the `else` clause if the loop is never entered:

```pycon
>>> for k, v in {}.items():
...     ...
... else:
...     print(k)
...
Traceback (most recent call last):
File "<python-input-0>", line 4, in <module>
  print(k)
        ^
NameError: name 'k' is not defined
```

This also applies to [unnecessary-list-index-lookup (PLR1736)](https://docs.astral.sh/ruff/rules/unnecessary-list-index-lookup/#unnecessary-list-index-lookup-plr1736):

```pycon
>>> for k, v in enumerate([]):
...     ...
... else:
...     print(k)
...
Traceback (most recent call last):
  File "<python-input-1>", line 4, in <module>
    print(k)
          ^
NameError: name 'k' is not defined
```

Although we currently flag this too: https://play.ruff.rs/da8e5ff1-0dba-4e59-a5c1-a84428e92819

The other class of issue I noticed was that we also emit diagnostics if the key or value variable is shadowed:

```py
def foo(d: dict):
    for k, v in d.items():
        for k in foo:
            print(d[k])  # not the same `k` as the `.items()` loop
        for v in foo:
            print(d[k])  # not the same `v` as the `.items()` loop
```

This only seems to be the case if the shadowing binding comes from a `for` loop, so I suspect we're checking the `BindingKind` but not the actual location of the binding. We should fix these too.

## Summary
- [ ] Fix `else` handling in `PLR1736`
- [ ] Fix shadowing in `PLR1733`
- [ ] Fix shadowing in `PLR1736`
